### PR TITLE
Optimizing menu item parent calculation

### PR DESCRIPTION
**Pull request by Hackwar**
mod_menu checks for every menu item if the item is a parent of another menu item. The method to do that via $menu->getItems() is pretty expensive and can create great delays on bigger sites. The proposed code will reduce this immensely by only checking those menu items that can have a child due to lft and rgt differing by more than one. Unfortunately that will not automatically mean that a menu item actually has children, since children could be present, but unpublished or in a different language. In those cases lft and rgt differ, but there is no child item anyway.

This will not change existing sites, but will mean a performance improvement. So the best way to test this would be to compare execution times on sites with a lot of menu items (>500). Another way would be a code review by a maintainer. :smile: 

This came up because of the insistence of @volandku in the Joomla IRC channel on Freenode who made me look into this. Thank you for being persistent. :smile:

### CODE REVIEW

**Comment by andrepereiradasilva** [comment]
I have tested this item :red_circle: unsuccessfully on 51de1ebadd0092cdfb27fab8dda132df4089d729<br /><br />I get this php notices:
```
Notice: Undefined property: stdClass::$rgt in /path/to/joomla-cms/modules/mod_menu/helper.php on line 77
Notice: Undefined property: stdClass::$lft in /path/to/joomla-cms/modules/mod_menu/helper.php on line 777
```<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@Hackwar, instead of doing this:
```php
foreach ($items as $i => $item)
{
	/* Code removed for simplifying */
	$item->deeper     = false;
	$item->shallower  = false;
	$item->level_diff = 0;
	
	if (isset($items[$lastitem]))
	{
		$items[$lastitem]->deeper     = ($item->level > $items[$lastitem]->level);
		$items[$lastitem]->shallower  = ($item->level < $items[$lastitem]->level);
		$items[$lastitem]->level_diff = ($items[$lastitem]->level - $item->level);
	}

	$item->parent = (boolean) ($item->rgt > $item->lft + 1) && $menu->getItems('parent_id', (int) $item->id, true);
	/* Code removed for simplifying */
}

if (isset($items[$lastitem]))
{
	$items[$lastitem]->deeper     = (($start?$start:1) > $items[$lastitem]->level);
	$items[$lastitem]->shallower  = (($start?$start:1) < $items[$lastitem]->level);
	$items[$lastitem]->level_diff = ($items[$lastitem]->level - ($start?$start:1));
}
```

Can't we do something like this:
```
foreach ($items as $i => $item)
{
	/* Code removed for simplifying */
	$item->deeper     = false;
	$item->shallower  = false;
	$item->level_diff = 0;
	$item->parent = 0;

	if (isset($items[$lastitem]))
	{
		$items[$lastitem]->deeper     = ($item->level > $items[$lastitem]->level);
		$items[$lastitem]->shallower  = ($item->level < $items[$lastitem]->level);
		$items[$lastitem]->level_diff = ($items[$lastitem]->level - $item->level);
		$items[$lastitem]->parent = ($items[$lastitem]->level_diff == -1);
	}
	/* Code removed for simplifying */
}

if (isset($items[$lastitem]))
{
	$items[$lastitem]->deeper     = (($start?$start:1) > $items[$lastitem]->level);
	$items[$lastitem]->shallower  = (($start?$start:1) < $items[$lastitem]->level);
	$items[$lastitem]->level_diff = ($items[$lastitem]->level - ($start?$start:1));
	$items[$lastitem]->parent = ($items[$lastitem]->level_diff === -1);
}
```

In other words, can't we just don't use the `getItems` to calculate if current menu item is a parent or not and use the already calculated level_diff to do that.

**Comment by Hackwar** [comment]
First of all, sorry for the broken code. Yes, you are right, your proposal is even better. I added your code. I'm setting the parent to false by default to have it set at all times.

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on df6acd259cb3062ad0ec1b1499f72aa6cea3e10e<br /><br />Thank you for find that needed performance optimization.
This can have some performance impact on sites with a lot of menu items.

Correct me if i'm wrong but the last element can't be the parent of anything, since it's the last.
That is why you didn't put the `$items[$lastitem]->parent     = ($items[$lastitem]->level_diff == -1);` in the setting of last item after the for cycle as i suggested.
Makes sense.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by efEris** [comment]
I have tested this PR :red_circle: unsuccessfully on df6acd2.

Simple test, I take default installation added 5 Menu Items:

* Item 1
  * Item 1a
* Item 2
  * Item 2a
    * Item 2b

Module with "Show Sub-menu Items" off shows on active "Item 1" a difference at the last item, in this case "Item 2".

Without patch: parent = true
With patch: parent = false

If "Show Sub-menu Items" is on the "Item 2" parent value is correct (true).

Also if "Show Sub-menu Items" is on the the complete testcase is correct.

**Comment by andrepereiradasilva** [comment]
@efEris we know the last item parent value can be not ok. See my comment when i tested.

> Correct me if i'm wrong but the last element can't be the parent of anything, since it's the last.
> That is why you didn't put the $items[$lastitem]->parent = ($items[$lastitem]->level_diff == -1); in the setting of last item after the for cycle as i suggested.
> Makes sense.

But in what that affected your test? Did the menu not appeared correctly?

**Comment by efEris** [comment]
@andrepereiradasilva sure because the li-tag for the last element doesn't get the "parent" class, that could lead to a visual problem if you use this class (the only way) to mark the link as parent item.

**Comment by andrepereiradasilva** [comment]
Ok, let's wait for @Hackwar to solve that.

**Comment by Hackwar** [comment]
Please test this again. This should solve the previous issues.

**Comment by efEris** [comment]
Hi,

while testing this I think I came to a better solution.

Add this to the beginning of the foreach loop before filtering the unseen items.
```
$item->parent = false;
if (isset($items[$lastitem])
  && $items[$lastitem]->id == $item->parent_id)
{
  $items[$lastitem]->parent = true;
}
```

btw. is it normal that the joomla patch tester overrides the complete file with a incompatible version?

**Comment by andrepereiradasilva** [comment]
I think the prior solution (before revert) works, only is needed to add this line after the for cycle to set the last parent https://github.com/Hackwar/joomla-cms/blob/df6acd259cb3062ad0ec1b1499f72aa6cea3e10e/modules/mod_menu/helper.php#L130
```
$items[$lastitem]->parent     = ($items[$lastitem]->level_diff == -1);
```

**Comment by efEris** [comment]
@andrepereiradasilva I tested this first, but no that will not work because we don't have the child at this point. if you can please try my solution.

**Comment by andrepereiradasilva** [comment]
ok i will check. thanks

**Comment by andrepereiradasilva** [comment]
@efEris i tried you solution and it worked. And without the `getItems`. Good work.

How did i test: 

- Use the current php file (before this PR) and removed this line: https://github.com/joomla/joomla-cms/blob/staging/modules/mod_menu/helper.php#L85.

- As you said, put your code in this line https://github.com/joomla/joomla-cms/blob/staging/modules/mod_menu/helper.php#L56, so it becomes
```php
foreach ($items as $i => $item)
{
	$item->parent = false;
	if (isset($items[$lastitem]) && $items[$lastitem]->id == $item->parent_id)
	{
		$items[$lastitem]->parent = true;
	}
	// code removed for brevity
}
```

@Hackwar can you check? With @efEris solution i think we can get rid of `getItems` and with that improving performance even further.

**Comment by brianteeman** [comment]
@hackwar any comment on the proposed improvement from @efEris would be good to move this PR forward if we can<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by roland-d** [comment]
@Hackwar Another nudge so we can move this PR forward.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by Hackwar** [comment]
I don't know which one is @efEris. So far this PR is complete the way it is.

**Comment by HLeithner** [comment]
@Hackwar sorry I renamed the github Username

Instead of trying to get the ->parent value based on the level_diff its better to set it at the beginning before we delete the information.

Add the following code to the beginning of the foreach loop at Line 57, see comment from @andrepereiradasilva

```php
    $item->parent = false;
    if (isset($items[$lastitem]) && $items[$lastitem]->id == $item->parent_id)
    {
        $items[$lastitem]->parent = true;
    }`

**Comment by Hackwar** [comment]
I've added the change like you proposed @HLeithner.

**Comment by HLeithner** [comment]
One thing, adding ", m.lft, m.rgt" to the query is no longer necessary.

File: modules/mod_menu/helper.php [hunk 1]
```diff
@@ -54,6 +54,12 @@ public static function getList(&$params)
 			{
 				foreach ($items as $i => $item)
 				{
+					$item->parent = false;
+					if (isset($items[$lastitem]) && $items[$lastitem]->id == $item->parent_id)
```
Comment (roland-d) [inline modules/mod_menu/helper.php]: CS: We need an empty line before the if statement.

**Comment by andrepereiradasilva** [comment]
> One thing, adding ", m.lft, m.rgt" to the query is no longer necessary.

**Comment by andrepereiradasilva** [comment]
Menu with 500+ items

Before patch
image

After patch
image

So apox half the time

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on f1aa5a66faed754284ff441bd46c34af7b3095d6<br /><br />as comment above<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by gwsdesk** [comment]
I have tested this item :white_check_mark: successfully on c4d5c8cd12e00ba010b5233952012126eb31c937<br /><br />I have tested this on a stock Joomla distro (3.6.3) will "Learning Sample data" in 4 languages and it has a visible faster loading time compared to pre-patch<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by laoneo** [comment]
This issue has two successful tests.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/joomla-cms/8863</a>.</sub>

**Comment by wilsonge** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8863">issues.joomla.org/tracker/joomla-cms/8863</a>.</sub>

**Comment by brianteeman** [comment]
Milestone changed to 3.7 as it is now not planned to have a 3.6.4 release

**Pull request merged**